src/manifold/deferred.clj

```diff
@@ -1538,141 +1545,109 @@
   [d f]
   (let [d' (deferred)]
     (on-realized d
                  (fn [val] (success! d' (f val)))
                  (fn [error] (error! d' error)))
     d'))
 
-(defn- then-apply [d ^Function f]
-  (assert-some f)
-  (fmap-deferred d #(.apply f %)))
+(defn- completion-stage-result [d f result]
+  (when result
+    (shallow-connect d result))
+  (shallow-chain (or result d) f))
 
-(def-async-for then-apply)
+(defn- then-apply [d ^Function f result]
+  (assert-some f)
+  (completion-stage-result d #(.apply f %) result))
 
-(defn- then-accept [d ^Consumer c]
+(defn- then-accept [d ^Consumer c result]
   (assert-some c)
-  (fmap-deferred d #(.accept c %)))
-
-(def-async-for then-accept)
+  (completion-stage-result d #(.accept c %) result))
 
-(defn- then-run [d ^Runnable f]
+(defn- then-run [d ^Runnable f result]
   (assert-some f)
-  (fmap-deferred d (fn [_] (.run f))))
-
-(def-async-for then-run)
+  (completion-stage-result d (fn [_] (.run f)) result))
 
-
-(defn- then-combine [d other ^BiFunction f]
+(defn- then-combine [d other ^BiFunction f result]
   (assert-some other f)
-  (fmap-deferred (zip d other)
-                (fn [[x y]] (.apply f x y))))
-
-(def-async-for-dual then-combine)
+  (completion-stage-result (zip d other)
+                           (fn [[x y]] (.apply f x y))
+                           result))
 
-
-(defn- then-accept-both [d other ^BiConsumer f]
+(defn- then-accept-both [d other ^BiConsumer f result]
   (assert-some other f)
-  (fmap-deferred (zip d other)
-                (fn [[x y]] (.accept f x y))))
-
-(def-async-for-dual then-accept-both)
+  (completion-stage-result (zip d other)
+                           (fn [[x y]] (.accept f x y))
+                           result))
 
-
-(defn- run-after-both [d other ^Runnable f]
+(defn- run-after-both [d other ^Runnable f result]
   (assert-some other f)
-  (fmap-deferred (zip d other)
-                (fn [[_ _]] (.run f))))
-
+  (completion-stage-result (zip d other)
+                           (fn [[_ _]] (.run f))
+                           result))
 
-(def-async-for-dual run-after-both)
-
-
-(defn- apply-to-either [d other ^Function f]
+(defn- apply-to-either [d other ^Function f result]
   (assert-some other f)
-  (then-apply (alt d other) f))
-
-(def-async-for-dual apply-to-either)
+  (then-apply (alt d other) f result))
 
-
-(defn- accept-either [d other ^Function f]
+(defn- accept-either [d other ^Function f result]
   (assert-some other f)
-  (then-accept (alt d other) f))
-
-(def-async-for-dual accept-either)
+  (then-accept (alt d other) f result))
 
-
-(defn- run-after-either [d other ^Function f]
+(defn- run-after-either [d other ^Function f result]
   (assert-some other f)
-  (then-run (alt d other) f))
-
-(def-async-for-dual run-after-either)
+  (then-run (alt d other) f result))
 
-
-(defn- then-compose [d ^Function f]
+(defn- then-compose [d ^Function f result]
   (assert-some f)
   (let [d' (deferred)]
```

Comment:
Is this still necessary? Shouldn't the provided target (result) be used instead?

Reply:
This is a bit subtle indeed :sweat_smile: The purpose of `d'` is to unwrap the deferred returned from `f` (= `fd`), so AFAICT we still need it. If it helps: `thenCompose` is basically the `bind` / `flatMap` operation of the `CompletionStage` monad :grimacing: